app/jobs/submissions_job.rb

```diff
@@ -5,6 +5,101 @@ def add_warning_messages(messages)
     Rails.logger.error msg
   end
 
+  def copy_old_grading_data(new_submission, grouping)
+    # at this point, it will only have one result
+    new_result = new_submission.current_result
+
+    # get the last submission that wasn't the one we just created
+    old_submission = grouping.submissions.where.not(id: new_submission.id)
+                             .order(created_at: :desc)
+                             .first
+
+    # no old submissions => this option does nothing
+    return if old_submission.nil?
+
+    old_result = old_submission.non_pr_results
+                               .where(remark_request_submitted_at: nil)
+                               .last
+
+    result_data = [
+      { old: old_result.annotations, new: new_result.annotations },
+      { old: old_result.marks, new: new_result.marks },
+      { old: old_result.extra_marks, new: new_result.extra_marks }
+    ]
+
+    # copy over data from old result to new result
+    result_data.each do |result_set|
+      # get rid of the existing empty records so we can replace them; some results
+      # have default models created, so this is necessary
+      result_set[:new].destroy_all
+
+      result_set[:old].each do |record|
+        record_dup = record.dup
+        record_dup.update(result_id: new_result.id)
+
+        add_warning_messages(record_dup.errors.full_messages) if record_dup.errors.present?
+      end
+    end
+
+    # amend submission file ids for annotations (new submission files are
+    # created for each submission)
+    new_result.reload.annotations.each do |annotation|
+      filename = annotation.submission_file.filename
+      path = annotation.submission_file.path
+
+      new_submission_file = new_submission.submission_files.where(filename: filename, path: path).first
```

Comment:
I wonder whether there is a better way to do this (rather than by filename and path).